wzprof.go

```diff
@@ -10,30 +11,116 @@ import (
 	"unsafe"
 
 	"github.com/google/pprof/profile"
+	"github.com/tetratelabs/wazero"
 	"github.com/tetratelabs/wazero/api"
 	"github.com/tetratelabs/wazero/experimental"
 	"golang.org/x/exp/slices"
 )
 
+type stackIteratorMaker interface {
+	Make(mod api.Module, def api.FunctionDefinition, wasmsi experimental.StackIterator) experimental.StackIterator
+}
+
+// Runtime contains information specific to the guest runtime.
+type Runtime struct {
```

Comment:
I'm concerned the `Runtime` name may be really confusing since we're depending on wazero which has a `Runtime` type, and we're also dealing with the Go runtime, etc... lots of overload!

Maybe an alternative model
```go
type Symbolizer interface {
    Locations(fn experimental.InternalFunction, pc experimental.ProgramCounter) (uint64, []location)
}

type SymbolTable struct {
  // things that the runtime has
}

func NewSymbolTable() *SymbolTable {
    ...
}

func (s *SymbolTable) Build(wasm []byte, module wazero.CompiledModule) error {
    ...
} 

// implements Symbolizer so it can be passed to the profiler constructors
func (s *SymbolTable) Locations(fn experimental.InternalFunction, pc experimental.ProgramCounter) (uint64, []location) {
     ...
}

func NewCPUProfiler(symbols Symbolizer, ...) *CPUProfiler {
    ...
}

func NewMemoryProfiler(symbols Symbolizer, ...) *MemoryProfiler {
    ...
}
```

What do you think?

Reply:
I agree with the "runtime" overload! I moved away from Symbol* because it also has to decide which method to use to walk the call stack, which seems beyond the realm of symbolization.

How about `GuestSupport`?

Reply:
Maybe call it `Program`?

Reply:
After trying it, seemed read weird to me because i kept thinking that the program was the one I was writing (main.go). The word that feels the best so far is `Support`:  https://github.com/stealthrocket/wzprof/pull/85/commits/24ba7460c37bbadad3cc9d46aec9335b4b878879

Reply:
Support feels like we're inventing a brand new concept in the world of profiling, which makes me think it may not communicate well what this type is meant to do. Now that could totally be me not being knowledgeable, did you have examples of how a similar type may have been named in other code bases?

Reply:
I don't know of other general-purpose profiling tools that adapt their behavior to what they are profiling. Maybe the closest would be Sentry, which seems to have the concept of ["platform"](https://docs.sentry.io/product/profiling/) – but they distribute an SDK for each.

Reply:
Maybe we need a native speaker to chime in @chriso

Reply:
How about we avoid new terms entirely? I might be missing context, but it appears that the existing `Support` (`/ Runtime`) type was added because the CPU profiler and memory profiler shared data (e.g. symbols) and code (e.g. symbolizer). How about a top-level abstract `Profiler` type that prepares the data, provides the helpers, and also acts as a factory for the concrete profilers?

```go
type Profiler struct {
   // shared data, built lazily
}

func (p *Profiler) sharedButPrivateMethod() 

func (p *Profiler) CPUProfiler(constructorArgs...) *CPUProfiler { ... }
func (p *Profiler) MemoryProfiler(constructorArgs...) *MemProfiler { ... }

type CPUProfiler struct {
  p *Profiler
}

type MemProfiler struct {
  p *Profiler
}
```

Usage:

```go
p := wzprof.NewProfiler(wasmCode)

cpu := p.CPUProfiler(wzprof.HostTime(prog.hostTime))
mem := p.MemoryProfiler(wzprof.InuseMemory(prog.inuseMemory))
```

With this approach, we can tightly control the API surface area by controlling what's exported. We can hide implementation details like symbols, and prevent misuse.

Reply:
works for me :)

Reply:
We already have the `Profiler` interface though. It's used by `Handler` to generate the HTTP handler for pprof. I've played with it a bit, and feel like the name `Profiling` works nicely:

```go
p := wzprof.ProfilingFor(wasmCode)
cpu := p.CPUProfiler()
mem := p.MemoryProfiler()
```

(https://github.com/stealthrocket/wzprof/pull/85/commits/6639ea56e9c89ab12ed042e37b42ef0f2401b7f0)

Reply:
Let's go with it, we're still pre-v1 so if we find a better name we can change it later 👍